paddle/operators/squared_l2_norm_op.h

```diff
@@ -0,0 +1,64 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+// Out = sum(square(X))
+template <typename Place, typename T>
+class SquaredL2NormKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &context) const override {
+    const framework::Tensor *X = context.Input<framework::Tensor>("X");
+    framework::Tensor *Out = context.Output<framework::Tensor>("Out");
+    Out->mutable_data<T>(context.GetPlace());
+
+    auto x = framework::EigenVector<T>::Flatten(*X);
+    auto out = framework::EigenVector<T>::Flatten(*Out);
+    auto place = context.GetEigenDevice<Place>();
+
+    out.device(place) = x.square().sum();
```

Comment:
1.  The `SquaredL2Norm` is good name, the square of L2 norm.

2. But if for L2 regularization, usually the overall cost doest not contain the regularization term (also called a weight decay term) in most framework, including the old PaddlePaddle framework.  So there is no need to use this op in the forward. Although, the cost function (as follows) contains this term.

    see the formula in this link : http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm

    And in the backward backpropagation,  [the derivative of the overall cost function  J(w, b)](http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm) is: 

    see the formula in this link : http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm

    the weight decay term will be a linear operation on W.  (Discussed with @lcy-seso serveral days ago).  So only needs a scale op for the L2 Regularization in the parameter updating process.  Also can see the momentum updating formula in the [paper](http://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks), the weight decay is a linear operation on W.

So I'm not sure whether this op is needed and whether there is another use scenarios.

Reply:
1. Regularization Loss is a term of the overall loss equation. If we want to plot an overall loss graph along every iteration for a classification task, the loss should equal `classification loss + regularization loss`.

2. `weight decay` just fits `L2-norm regularization`. We try to implement a common way for regularization, which fits both L1 and L2. 

3. We will provide a `weight decay` in the future PRs. It is not a conflict that we provide both `weight decay` and `L2 Norm` operator.

Reply:
@qingqing01 @lcy-seso Both these regularization techniques have their advantages and disadvantages. Let me summarize them as follows:
 
| Weight Decay in Optimizer 	| Separate Operators for regularization 	|
|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|----------------------------------------------------------------------------------------------------------:	|
| No Forward prop op, hence faster 	| There will be a forward prop op 	|
| Will not support making plots of total loss vs epoch/iteration 	| Will support those plots 	|
| This does not generalize well beyond L2 and L1 regularization. Use of Batch norm and layer norm might invalidate this approach 	| This is a very general approach and can be applied to any  kind of network 	|
| It is not easy for researchers to add new regularizers in this framework because the regularization is tightly coupled with optimizers. They might  have to change all optimizers. 	| Adding new regularization schemes is easy as the code for regularization  is independent of optimization. 	|
| Frameworks that support this: Pytorch, Caffe 	| Frameworks that support this: Tensorflow, Theano, Lasagne 	|

Reply:
I think that forward prop op might be necessary because while training the model, it is a very common practice to check convergence by making plots of the total loss function vs time. Also during inference, an intelligent executor can easily prune out the graph to remove the regularization nodes. 

Also I agree with @reyoung that we can also implement the weight decay separately. That can be only for the case of the L2 and L1 Penalty loss. 

Please let me know what you think about this plan?

Reply:
- For the plotting:
In the old PaddlePaddle framework and Caffe, I think we usually plot the loss with regularization term vs time.

- For the regularization

I always agree to separate operators for regularization, not implementing in the optimizer.  But I think whether to add the regularization loss to the overall loss depends on the users, not by default. Since it all add more calculation during the training. And the default regularization is only to use `L2LrRegularizer` (a linear operator)/ `L1LrRegularizer` operators like the implementation in https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/parameter/Regularizer.h for the parameters updating. 

- About this PR
I agree to merge this Op. But whether to use it in the forward and add the regularization loss to the overall loss depends on the users, not default by the framework, if using the regularization in the training.

Reply:
@qingqing01 @lcy-seso Thank you for going through the PR. I discussed this today with @reyoung and we feel that your point is valid. We can add these ops but whether to use them in forward and add to the loss, will be the user's choice. In another PR, I will add separate operators for regularization which will be used only in the backward pass and not implemented in the optimizer. 
In case of L2, this will be a simple scale op and in the case of L1 regularization this will be a combination of scale and sign op. 
Thank you so much for your feedback on this.